### ISSUE TITLE
AMS doesn't serialize params

### ISSUE THREAD

**Issue by ivanjolic95** [rails-api/active_model_serializers#1595]
A few months ago I was using AMS JSON API adapter with ember and everything was working well.

Pulling the data from the API seems to be working, but when I try to save something from my ember app I get following errors:

`"#<ActionController::ParameterMissing: param is missing or the value is empty: itinerary>"`
`"#<ActionController::ParameterMissing: param is missing or the value is empty: itinerary_item>"`
```"#<NoMethodError: undefined method `itinerary_item_url' for #<Api::V1::ItineraryItemsController:0x000000069a4240>>"```

#### Steps to reproduce
I've followed [ember and json api integration](https://github.com/rails-api/active_model_serializers/blob/master/docs/integrations/ember-and-json-api.md) and added `ActiveModel::Serializer.config.adapter = :json_api` to an initializer. Then I created serializers like always, added controllers that are based on `ActiveController::API` and added routes.

Here's my ItinerariesController:

```
class Api::V1::ItinerariesController < ActionController::API

  def index
    load_itineraries
    render json: @itineraries
  end

  def show
    load_itinerary
    render json: @itinerary
  end

  def create
    build_itinerary
    save_itinerary(status: :created) or render_error
  end

  def update
    load_itinerary
    build_itinerary
    save_itinerary(status: nil) or render_error
  end

  def destroy
    load_itinerary
    @itinerary.destroy
  end

  private
    
    def load_itineraries
      @itineraries = itinerary_scope
    end

    def load_itinerary
      @itinerary = itinerary_scope.find(params[:id])
    end

    def build_itinerary
      @itinerary ||= itinerary_scope.build
      @itinerary.attributes = itinerary_params
    end

    def save_itinerary(params)
      if @itinerary.save
        render json: @itinerary, status: params[:status], location: @itinerary
      else
        false
      end
    end

    def itinerary_params
      itinerary_params = params.require(:itinerary)
        .permit(
          :tour_id, :tour_request_id
        )
    end

    def itinerary_scope
      Itinerary.filtered(params[:filter])
    end

    def render_error
      render json: @itinerary.errors, status: :unprocessable_entity
    end

end
```

#### Environment

* AMS master branch
* rails Rails 5.0.0.beta3
* ember 2.3.0-beta.2
* gem ember-cli-rails 0.7.1

**Comment by NullVoxPopuli** [comment]
Hi, I use ember with rails. The way I save things is by deconstructing the json-api format, like so:

```ruby
 def update_package_params
    params
      .require(:data)
      .require(:attributes)
      .permit(
        :name,
        :initial_price, :at_the_door_price,
        :attendee_limit, :expires_at, :requires_track,
        :ignore_pricing_tiers, :deleted_at
      )
  end
```
see: 
https://github.com/NullVoxPopuli/aeonvera/blob/master/app/controllers/api/packages_controller.rb

**Comment by beauby** [comment]
If your client sends a JSON API payload, then the following
```ruby
itinerary_params = params.require(:itinerary)
```
should indeed fail because your payload would look like:
```ruby
{
  data: {
    type: 'itineraries',
    attributes: {
      tour_id: '1',
      tour_request_id: '1'
    }
  }
}
```
(notice the top-level `data` property).

Moreover, you should probably avoid exposing foreign keys as attributes (since they represent a relationship).

TL;DR: do
```ruby
```ruby
itinerary_params = params.require(:data).permit(attributes: [:tour_id, :tour_request_id])
```
```

**Comment by NullVoxPopuli** [comment]
I don't have a super elegant way to handle relationships -- here is what I currently do:

```ruby
  def create_membership_renewal_params
    data = params.require(:data)
    attributes = data
      .require(:attributes)
      .permit(:start_date)

    relationships = data.require(:relationships)

    user = relationships
      .require(:member)
      .require(:data)
      .permit(:id)

    membership_option = relationships
      .require(:membership_option)
      .require(:data)
      .permit(:id)

    attributes.merge(
      user_id: user[:id],
      membership_option_id: membership_option[:id])
  end
```

**Comment by beauby** [comment]
Note that you could both benefit from the deserialization methods that AMS provides.

**Comment by NullVoxPopuli** [comment]
AMS already has that? I don't see it documented anywhere? (granted, I didn't open all the docs...)

**Comment by remear** [comment]
@NullVoxPopuli https://github.com/rails-api/active_model_serializers/pull/1431

**Comment by ivanjolic95** [comment]
@remear is it usable? When I tried it on my data all I got back was an empty hash?

**Comment by NullVoxPopuli** [comment]
@remear legit, we should get those docs merged

**Comment by ivanjolic95** [comment]
Yep, it works, it didn't work in console :+1: thanks everyone!
